server/storage/sqldb.go

```diff
@@ -45,13 +45,44 @@ func translateOldVersionError(err error) error {
 	return err
 }
 
+// TufFilesInChannels scopes gorm queries to a particular channel or set of channels
+func TufFilesInChannels(channels ...*Channel) func(db *gorm.DB) *gorm.DB {
+	channelIDs := []uint{}
+	for _, channel := range channels {
+		channelIDs = append(channelIDs, channel.ID)
+	}
+
+	// If no channels were added, check Published
+	if len(channelIDs) == 0 {
+		channelIDs = []uint{Published.ID}
+	}
+
+	// If only one channel is provided, we don't need to group/collect
+	if len(channelIDs) == 1 {
+		return func(db *gorm.DB) *gorm.DB {
+			return db.Joins("INNER JOIN channels_tuf_files ON tuf_files.id = channels_tuf_files.tuf_file_id").
+				Where("channel_id = ?", channelIDs[0])
+		}
+	}
+
+	// Find tuf files matching all passed channels
+	return func(db *gorm.DB) *gorm.DB {
+		return db.Joins("INNER JOIN channels_tuf_files ON tuf_files.id = channels_tuf_files.tuf_file_id").
+			Where("channel_id IN (?)", channelIDs).
```

Comment:
Sorry, gorm makes this harder to read than it should be.

The scope says:

 - Ignore any roles that aren't in the set of requested channels
 - Count how many distinct channels a tuf file is in 
 - Because (`tuf_file_id`, `channel_id`) is unique, if number of unique channels == number of requested channels, the tuf file is in all requested channels.

For your specific example it would be something like this:

- Query by `GUN, TS` in `Alice` and `Carol` channels
- `tuf_files` is joined with `channel_tuf_files`
- All rows in the joined table with `channel_id` != `Alice` or `Carol` are discarded
  - this is the part that specifically answers your question - there is no longer a (`5`, `GUN`, `Timestamp`, `Bob`) row in the result set
- The result set only has (`5`, `Carol`)
- `COUNT` of that set is 1
- 2 channels were requested, so the result set returns empty